### ISSUE TITLE
Elementor Popup is NOT keyboard accessible

### ISSUE THREAD

**Issue by kelsS** [elementor/elementor#9788]
## Prerequisites

- [X] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [X] The issue still exists against the latest stable version of Elementor.

## Description

I attempted to use a keyboard to navigate through an Elementor popup window and discovered that it is not fully keyboard accessible and is a keyboard trap.

## Steps to reproduce
1. Using the keyboard click on the button that activates an Elementor popup
2. Tab through the form fields then try and close the popup using the keyboard only.
3. There is no way to close a popup using the keyboard only which creates a keyboard trap for the end user. Expected the popup to be fully keyboard accessible.

## Isolating the problem

- [X] This bug happens with only Elementor plugin active (and Elementor Pro).
- [X] This bug happens with a default WordPress theme active.
- [X] I can reproduce this bug consistently using the steps above.

## Environment

<details>
<summary>System Info</summary>

```bash
== Server Environment ==
	Operating System: Linux
	Software: Apache
	MySQL version: Percona Server (GPL), Release 84.0, Revision 47234b3 v5.6.40-84.0
	PHP Version: 7.1.33
	PHP Max Input Vars: 3000
	PHP Max Post Size: 128M
	GD Installed: Yes
	ZIP Installed: Yes
	Write Permissions: All right
	Elementor Library: Connected

== WordPress Environment ==
	Version: 5.3
	WP Multisite: Yes
	Max Upload Size: 1 MB
	Memory limit: 768M
	Permalink Structure: /%postname%/
	Language: en-US
	Timezone: 0
	Debug Mode: Inactive

== Theme ==
	Name: WP Bootstrap Starter
	Version: 3.3.2
	Author: Afterimage Designs
	Child Theme: No

== User ==
	User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:70.0) Gecko/20100101 Firefox/70.0

== Active Plugins ==
	Classic Editor
		Version: 1.5

	Elementor
		Version: 2.7.5
		Author: Elementor.com

	Elementor Pro
		Version: 2.7.3
		Author: Elementor.com
```
</details>

**Comment by swinggraphics** [comment]
Furthermore, triggering a popup does not move focus to the popup.

**Comment by ContactDigital** [comment]
Is there any updates on this and whether it's being addressed? It would be @kelsS if you could label this under Accessibility as I couldn't see anything regarding pops (an issue we currently have now), so it may help bring more awareness.

**Comment by imaximon** [comment]
+1

**Comment by libbeydesign** [comment]
Following. Any updates on this would be appreciated.

**Comment by jmgdesign** [comment]
+1. This is a big problem at the moment and one of several reasons why building with Elementor is problematic for creating an accessible website.

**Comment by tbusillo** [comment]
I don't know this is technically a keyboard trap. I just tested it out and the popups (at least the one I tested) _do_ respond to pressing the "escape" key. 

WCAG 2.1.2 states, "If keyboard focus can be moved to a component of the page using a keyboard interface, then **focus can be moved away from that component using only a keyboard interface**, and, if it requires more than unmodified arrow or tab keys or other standard exit methods, the user is advised of the method for moving focus away. (Level A)". 

My read on that is you do not necessarily have to make a cancel/close button accessible, so long as you provide another means for exiting the modal. It looks like you may have to advise the user of the alternative method for closing the modal as well. 

Nonetheless, there is a relatively simple stop-gap solution for this (in case anyone needs it). Add this within an existing JS file, or even inline, just before </body>. 

```
jQuery(document).ready(function($){	
	setTimeout(function() {
		$(".dialog-close-button").attr('tabindex', 0);
		$(".dialog-close-button").on('keypress', function(e) {
			e.preventDefault();
			if ( e.which == 13 || e.which == 32 ) {
				$(this).trigger("click");
			}
		})
	}, 300);
}); 
```

The above does the following:

- Adds the tabindex attribute on the close button and sets it to '0'
- Adds an event listener that will trigger the .click() event listener already on the close button. The listener looks for a keypress of the space bar or return key. 

_Note: you have to use setTimeout() due to the dialog being programatically added to the page after the page loads. I couldn't figure out a way to make it not be the last element prior to </body>._

**Comment by adamr415** [comment]
+1 these accessibility issues should be a big concern as anyone building a site without them could be open to ADA lawsuits which are becoming very popular

**Comment by dwamgo** [comment]
Huge obstacle for WCAG compliance

**Comment by Bovelett** [comment]
It would be so great if you could fix this! :-)

**Comment by xander-13** [comment]
+1

**Comment by Simpa77** [comment]
+1

**Comment by heroicP** [comment]
+1

**Comment by JboyJW** [comment]
+1

**Comment by JboyJW** [comment]
I've elaborated on @tbusillo's solution by capturing it all from the click of the button that triggers the popup since the popup html doesn't get generated until the trigger is clicked.  I've also added additional attributes to make the popup and close button more accessible and added functionality to set the focus back to the clicked button when the popup is closed.  Keyboard and screen reader users rejoice!

Basic explanations are commented directly in the code:
```
jQuery(document).ready(function($){

  //Fix Elementor popup focus (need to add a "elementor-popup-button" class to the widget that triggers the popup first)
  $('.has-elementor-popup-button a').on('click',function(e){

    //Note the button that is clicked to go back to it after the popup is closed
    var $buttonClicked = $(this);

    //Wait for the popup to be injected into the page
    setTimeout(function() {

      //Make the dialog accessible and set focus to it
      $('.dialog-widget').attr('tabindex', 0).attr('role','dialog').attr('aria-modal','true').focus();

      //Listen for the Esc key so that the focus goes back to the clicked button when pressed
      $(window).on('keydown', function(ev) {
        var keycode = (ev.keyCode ? ev.keyCode : ev.which);
        if (keycode == 27) {
          $buttonClicked.focus();
        }
      });

      //Make the popup's close button accessible and set focus back to clicked button when clicked
      $(".dialog-close-button").attr('tabindex', 0).attr('role','button').attr('aria-label','Close');
      $(".dialog-close-button").on('click', function(ev) {
        $buttonClicked.focus();
      });

      //Trigger the close button to be clicked when Enter or Space is pressed on it
      $(".dialog-close-button").on('keypress', function(ev) {
        var keycode = (ev.keyCode ? ev.keyCode : ev.which);
        if ( keycode == 13 || keycode == 32 ) {
          $(this).trigger("click");
        }
      });

    }, 300);

  });
});
```

**Comment by mcartoon** [comment]
@JboyJW  Thank you for this script!
works fine for me

**Comment by JeffreyHester** [comment]
How is JBoyJW's solution best implemented? We are using Hello+Elementor Pro. Would love to have a tutorial.

**Comment by JboyJW** [comment]
@JeffreyHester @danielwstrategies There are many ways to add javascript to your site. Most themes have a place to add Custom JS, there are plugins, or you can add a .js file in your child theme.

I was using the Bridge theme, so I was able to add the code in Qode Options -> General -> Custom JS.

Your theme likely has general settings to add javascript or you can get a plugin like https://wordpress.org/plugins/insert-headers-and-footers/, go to the Scripts in Footer section, and type <script> then the code above and then </script>

**Comment by JeffreyHester** [comment]
@JboyJW I am also used to using Bridge, but we switch to Hello them recently due to Google speed.  I just tried this as suggested and it's not putting the focus in the first field in the form on the popup. Once you click there you can tab through. Ideas? You can see this in action at HesterDesignsDemo2.com when you click the top-right Online Banking button. Please advise.

**Comment by JboyJW** [comment]
@JeffreyHester Yes, we've been using Hello now on all recent projects for the same reason. Bridge is really bloated.

And yes, the script I provided sets focus to the popup itself as the popup I had didn't have a form on it.  To set focus to the first form field you'll need to update this line part of the code:
```
      //Make the dialog accessible and set focus to it
      $('.dialog-widget').attr('tabindex', 0).attr('role','dialog').attr('aria-modal','true').focus();
```
To be this instead:
```
      //Make the dialog accessible and set focus to the first form field by id
      $('.dialog-widget').attr('role','dialog').attr('aria-modal','true');
      $('#form-field-AXS01').focus();
```
There's probably a better way to do that so that it always gets the first form field regardless of what id it is, but that'll do for your purposes.

**Comment by JeffreyHester** [comment]
I am really appreciating your help @JboyJW. I must be doing something wrong, as this isn't working for me. I did title the first field on the form AXS01 just in case. Still no luck.

**Comment by tartinegithub** [comment]
@JboyJW, your solution worked great for me except when I'm using the dialog lightbox that uses a font-icon for the close button like this:
<div class="dialog-close-button dialog-lightbox-close-button">
<i class="eicon-close"></i>
</div>
And advice on how to get it to work here?
Thanks!

**Comment by kelly-childs** [comment]
@JboyJW, your solution worked great for me except when I'm using the dialog lightbox that uses a font-icon for the close button like this:

And advice on how to get it to work here? Thanks!

**Comment by diapproach** [comment]
> I've elaborated on @tbusillo's workaround by capturing it all from the click of the button that triggers the popup since the popup html doesn't get generated until the trigger is clicked. I've also added additional attributes to make the popup and close button more accessible and added functionality to set the focus back to the clicked button when the popup is closed. Keyboard and screen reader users rejoice!
> 
> Basic explanations are commented directly in the code:
> 
> ```
> jQuery(document).ready(function($){
> 
>   //Fix Elementor popup focus (need to add a "elementor-popup-button" class to the widget that triggers the popup first)
>   $('.has-elementor-popup-button a').on('click',function(e){
> 
>     //Note the button that is clicked to go back to it after the popup is closed
>     var $buttonClicked = $(this);
> 
>     //Wait for the popup to be injected into the page
>     setTimeout(function() {
> 
>       //Make the dialog accessible and set focus to it
>       $('.dialog-widget').attr('tabindex', 0).attr('role','dialog').attr('aria-modal','true').focus();
> 
>       //Listen for the Esc key so that the focus goes back to the clicked button when pressed
>       $(window).on('keydown', function(ev) {
>         var keycode = (ev.keyCode ? ev.keyCode : ev.which);
>         if (keycode == 27) {
>           $buttonClicked.focus();
>         }
>       });
> 
>       //Make the popup's close button accessible and set focus back to clicked button when clicked
>       $(".dialog-close-button").attr('tabindex', 0).attr('role','button').attr('aria-label','Close');
>       $(".dialog-close-button").on('click', function(ev) {
>         $buttonClicked.focus();
>       });
> 
>       //Trigger the close button to be clicked when Enter or Space is pressed on it
>       $(".dialog-close-button").on('keypress', function(ev) {
>         var keycode = (ev.keyCode ? ev.keyCode : ev.which);
>         if ( keycode == 13 || keycode == 32 ) {
>           $(this).trigger("click");
>         }
>       });
> 
>     }, 300);
> 
>   });
> });
> ```

thanks man!!!!! you are the man!!!!

**Comment by Sindhumanjuanth** [comment]
@diapproach your code works fine for me, but after reaching close button if I press tab, focus is going out of the website. But I want to make it to loop within popup, so please help me out

**Comment by Sindhumanjuanth** [comment]
> @diapproach your code works fine for me, but after reaching close button if I press tab, focus is going out of the website. But I want to make it to loop within popup, so please help me out

**Comment by ruben3022** [comment]
> > @diapproach your code works fine for me, but after reaching close button if I press tab, focus is going out of the website. But I want to make it to loop within popup, so please help me out

@diapproach @Sindhumanjuanth 
I Have the same issue! 
If anyone knows of a fix I would love to hear it.

**Comment by ruben3022** [comment]
> > @diapproach your code works fine for me, but after reaching close button if I press tab, focus is going out of the website. But I want to make it to loop within popup, so please help me out

@JboyJW @tbusillo 
I Have the same issue! 
If anyone knows of a fix I would love to hear it.

**Comment by kelly-childs** [comment]
@ruben3022 @JboyJW 
This usually works for me to trap the focus inside the popup until the user dismisses it:

`// add all the elements inside modal which you want to make focusable
const  focusableElements =
    'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])';
const modal = document.querySelector('#modal'); // select the modal by it's id

const firstFocusableElement = modal.querySelectorAll(focusableElements)[0]; // get first element to be focused inside modal
const focusableContent = modal.querySelectorAll(focusableElements);
const lastFocusableElement = focusableContent[focusableContent.length - 1]; // get last element to be focused inside modal


document.addEventListener('keydown', function(e) {
  let isTabPressed = e.key === 'Tab' || e.keyCode === 9;

  if (!isTabPressed) {
    return;
  }

  if (e.shiftKey) { // if shift key pressed for shift + tab combination
    if (document.activeElement === firstFocusableElement) {
      lastFocusableElement.focus(); // add focus for the last focusable element
      e.preventDefault();
    }
  } else { // if tab key is pressed
    if (document.activeElement === lastFocusableElement) { // if focused has reached to last focusable element then focus first focusable element after pressing tab
      firstFocusableElement.focus(); // add focus for the first focusable element
      e.preventDefault();
    }
  }
});

firstFocusableElement.focus();`

**Comment by ruben3022** [comment]
> @ruben3022 @JboyJW This usually works for me to trap the focus inside the popup until the user dismisses it:

Thanks a lot! it works great!

**Comment by ruben3022** [comment]
@kelly-childs Thanks a lot! it works great!

However when using a button to toggle open and close the popup it is not within the dom of the modal, is it possible to include it in the keyboard controls when in the modal?

in my case:
I have a header with a button that toggles a popup modal for the menu. I can open and close the popup with said button. 
It works great when closing using esc and re-highlighting the menu button, but when cycling through the popup modal the close button is not accesible.

If you have any idea i would love to hear!

**Comment by swinggraphics** [comment]
> when cycling through the popup modal the close button is not accessible.

Yet another aspect of how inaccessible pop ups are! The close button should be a `<button>`. Elementor decided to fight against basic HTML, best practices, accessibility standards, etc, and use an un-focusable div. >:[

**Comment by kelly-childs** [comment]
@swinggraphics is correct. Pseudo elements are not focusable so the close "button" needs to be an actual HMTL <button>. This will natively allow it to have keyboard focus. Add an aria-label to the <button> that describes its function. Ex: aria-label="close modal"

**Comment by nicholaszein** [comment]
Hello everyone!

We have great news! 🙌 

### 📢 We're happy to announce that the issue you raised was resolved in Elementor Pro **v3.11**! 🥳 

✔️ Feel free to check it out and update your plugin to the new version!

Cheers 🥂
